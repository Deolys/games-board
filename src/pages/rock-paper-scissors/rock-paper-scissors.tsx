import { useState } from "react";
import { Box, Button, Paper, Typography } from "@mui/material";
import rockIcon from "../../assets/icons/rock.svg";
import paperIcon from "../../assets/icons/paper.svg";
import scissorsIcon from "../../assets/icons/scissors.svg";
import { Header } from "../../components/header";

interface Choice {
  name: string;
  image: string;
}

const choices: Choice[] = [
  { name: "камень", image: rockIcon },
  { name: "ножницы", image: scissorsIcon}, 
  { name: "бумага", image: paperIcon},
];

export function RockPaperScissors() {
  const [userChoice, setUserChoice] = useState<Choice | null>(null);
  const [computerChoice, setComputerChoice] = useState<Choice | null>(null);
  const [userScore, setUserScore] = useState<number>(0);
  const [computerScore, setComputerScore] = useState<number>(0);
  const [result, setResult] = useState<string>("");

  const play = (choice: Choice) => {
    const computerChoice = choices[Math.floor(Math.random() * choices.length)];
    setUserChoice(choice);
    setComputerChoice(computerChoice);
    determineWinner(choice.name, computerChoice.name);
  };

  const determineWinner = (user: string, computer: string) => {
    if (user === computer) {
      setResult("Ничья!");
    } else if (
      (user === "камень" && computer === "ножницы") ||
      (user === "ножницы" && computer === "бумага") ||
      (user === "бумага" && computer === "камень")
    ) {
      setResult("Вы выиграли!");
      setUserScore((prev) => prev + 1);
    } else {
      setResult("Вы проиграли!");
      setComputerScore((prev) => prev + 1);
    }
  };

  return (
    <>
      <Header />
    <Paper style={{ padding: "20px", maxWidth: "500px", margin: "auto" }}>
      <Typography variant="h4" align="center">
        Камень, Ножницы, Бумага
      </Typography>
      <Box sx={{ display: "flex", justifyContent: "space-between", minHeight: "180px", mt: 3}}>
      <Box sx={{display: "flex", flexDirection: "column"}}>
        <Typography variant="h5" align="center">
          Ваш выбор:
        </Typography>
        <Box sx={{ display: "flex", justifyContent: "center", my: 3}}>
        {userChoice && <img src={userChoice.image} alt={userChoice.name}/>}
        </Box>
      </Box>
      <Box sx={{display: "flex", flexDirection: "column"}}>
      <Typography variant="h5" align="center">
        Выбор ИИ:
      </Typography>
      <Box sx={{ display: "flex", justifyContent: "center", my: 3}}>
      {computerChoice && <img src={computerChoice.image} alt={computerChoice.name}/>}
      </Box>
      </Box>
      </Box>
      <Typography
        variant="h5"
        align="center"
        style={{ color: result.includes("выиграли") ? "green" : "red" }}
      >
        Результат: {result}
      </Typography>
      <Box sx={{ display: "flex", justifyContent: "space-between"}}>
        <Typography variant="h6" align="center">
          Ваш счёт: {userScore}
        </Typography>
        <Typography variant="h6" align="center">
          Счёт ИИ: {computerScore}
        </Typography>
      </Box>
      <Box
        sx={{ display: "flex", justifyContent: "center", margin: "20px 0" }}
      >
        {choices.map((choice) => (
            <Button
            variant="contained"
            onClick={() => play(choice)}
            style={{ margin: "0 10px" }}
            key={choice.name}
          >
            {choice.name}
          </Button>
        ))}
      </Box>
    </Paper>
    </>
  );
}

export default RockPaperScissors;
