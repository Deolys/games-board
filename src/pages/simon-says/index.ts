export { SimonSaysGame } from "./simon-says";