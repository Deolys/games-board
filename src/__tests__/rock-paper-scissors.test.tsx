import { fireEvent, render, screen } from "@testing-library/react";
import { RockPaperScissors } from "../pages/rock-paper-scissors";
import '@testing-library/jest-dom';

describe('RockPaperScissors', () => {
  it('RockPaperScissors mounts properly', () => {
    const wrapper = render(<RockPaperScissors />);
    expect(wrapper).toMatchSnapshot();
  })

  test.each([
    ['камень', /Результат: (Вы выиграли!|Вы проиграли!|Ничья!)/i],
    ['ножницы', /Результат: (Вы выиграли!|Вы проиграли!|Ничья!)/i],
    ['бумага', /Результат: (Вы выиграли!|Вы проиграли!|Ничья!)/i],
  ])('При нажатии на кнопку результат должен быть определен', (buttonName, expectedResult) => {
    render(<RockPaperScissors />);
    const button = screen.getByRole('button', { name: new RegExp(buttonName, 'i') });

    fireEvent.click(button);

    const resultText = screen.getByText(expectedResult);
    expect(resultText).toBeInTheDocument();
  });
});