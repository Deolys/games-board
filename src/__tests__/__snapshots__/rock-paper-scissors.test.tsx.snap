// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RockPaperScissors RockPaperScissors mounts properly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1413kmy-MuiPaper-root-MuiAppBar-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-u4fbkj-MuiToolbar-root"
        >
          <a
            class="css-3qnz2e"
            href="/"
          >
            <img
              alt="Games board"
            />
            <p
              class="MuiTypography-root MuiTypography-body1 css-1waifce-MuiTypography-root"
            >
              Games board
            </p>
          </a>
          <div
            class="MuiStack-root css-1dzrg60-MuiStack-root"
          >
            <a
              class="MuiTypography-root MuiTypography-body1 css-1v5m6us-MuiTypography-root"
              href="/"
            >
              HOME
            </a>
            <button
              aria-haspopup="true"
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-17ovnxk-MuiButtonBase-root-MuiButton-root"
              id="composition-button"
              tabindex="0"
              type="button"
            >
              Games
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </header>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
        style="padding: 20px; max-width: 500px; margin: auto;"
      >
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter css-149xwtl-MuiTypography-root"
        >
          Камень, Ножницы, Бумага
        </h4>
        <div
          class="MuiBox-root css-1917ghf"
        >
          <div
            class="MuiBox-root css-j7qwjs"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
            >
              Ваш выбор:
            </h5>
            <div
              class="MuiBox-root css-j3ycse"
            />
          </div>
          <div
            class="MuiBox-root css-j7qwjs"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
            >
              Выбор ИИ:
            </h5>
            <div
              class="MuiBox-root css-j3ycse"
            />
          </div>
        </div>
        <h5
          class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
          style="color: red;"
        >
          Результат: 
        </h5>
        <div
          class="MuiBox-root css-gg4vpm"
        >
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-1kgxqm0-MuiTypography-root"
          >
            Ваш счёт: 
            0
          </h6>
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-1kgxqm0-MuiTypography-root"
          >
            Счёт ИИ: 
            0
          </h6>
        </div>
        <div
          class="MuiBox-root css-dnzamn"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
            style="margin: 0px 10px;"
            tabindex="0"
            type="button"
          >
            камень
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
            style="margin: 0px 10px;"
            tabindex="0"
            type="button"
          >
            ножницы
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
            style="margin: 0px 10px;"
            tabindex="0"
            type="button"
          >
            бумага
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-1413kmy-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-u4fbkj-MuiToolbar-root"
      >
        <a
          class="css-3qnz2e"
          href="/"
        >
          <img
            alt="Games board"
          />
          <p
            class="MuiTypography-root MuiTypography-body1 css-1waifce-MuiTypography-root"
          >
            Games board
          </p>
        </a>
        <div
          class="MuiStack-root css-1dzrg60-MuiStack-root"
        >
          <a
            class="MuiTypography-root MuiTypography-body1 css-1v5m6us-MuiTypography-root"
            href="/"
          >
            HOME
          </a>
          <button
            aria-haspopup="true"
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-17ovnxk-MuiButtonBase-root-MuiButton-root"
            id="composition-button"
            tabindex="0"
            type="button"
          >
            Games
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </header>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
      style="padding: 20px; max-width: 500px; margin: auto;"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter css-149xwtl-MuiTypography-root"
      >
        Камень, Ножницы, Бумага
      </h4>
      <div
        class="MuiBox-root css-1917ghf"
      >
        <div
          class="MuiBox-root css-j7qwjs"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
          >
            Ваш выбор:
          </h5>
          <div
            class="MuiBox-root css-j3ycse"
          />
        </div>
        <div
          class="MuiBox-root css-j7qwjs"
        >
          <h5
            class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
          >
            Выбор ИИ:
          </h5>
          <div
            class="MuiBox-root css-j3ycse"
          />
        </div>
      </div>
      <h5
        class="MuiTypography-root MuiTypography-h5 MuiTypography-alignCenter css-ph2xsj-MuiTypography-root"
        style="color: red;"
      >
        Результат: 
      </h5>
      <div
        class="MuiBox-root css-gg4vpm"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-1kgxqm0-MuiTypography-root"
        >
          Ваш счёт: 
          0
        </h6>
        <h6
          class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-1kgxqm0-MuiTypography-root"
        >
          Счёт ИИ: 
          0
        </h6>
      </div>
      <div
        class="MuiBox-root css-dnzamn"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
          style="margin: 0px 10px;"
          tabindex="0"
          type="button"
        >
          камень
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
          style="margin: 0px 10px;"
          tabindex="0"
          type="button"
        >
          ножницы
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
          style="margin: 0px 10px;"
          tabindex="0"
          type="button"
        >
          бумага
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
